File: src/gui/objects/textfield/autocomplete/ConfigSearch.java
package gui.objects.textfield.autocomplete;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;
import java.util.stream.Stream;

import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Node;
import javafx.scene.control.*;
import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyEvent;
import javafx.scene.layout.HBox;
import javafx.scene.layout.StackPane;
import javafx.scene.text.TextAlignment;

import gui.itemnode.ConfigField;
import gui.objects.textfield.autocomplete.ConfigSearch.Entry;
import gui.objects.window.stage.UiContext;
import layout.widget.WidgetFactory;
import util.action.Action;
import util.conf.Config;
import util.functional.Functors.Ƒ0;

import static main.App.Build.appTooltip;
import static util.Util.containsNoCase;
import static util.functional.Util.by;
import static util.functional.Util.stream;
import static util.graphics.Util.layStack;
import static util.graphics.Util.setMinPrefMaxSize;

/**
 *
 *
 * @author Martin Polakovic
 */
public class ConfigSearch extends AutoCompletion<Entry> {

    private final TextField textField;
    private final History history;
    private boolean ignoreEvent = false;

	@SafeVarargs
    public ConfigSearch(TextField textField, Supplier<Stream<Entry>>...searchTargets) {
        this(textField, new History(), searchTargets);
    }

	@SafeVarargs
    public ConfigSearch(TextField textField, History history, Supplier<Stream<Entry>>... searchTargets) {
        super(
        	textField,
	        s -> {
        		String text = s.getUserText();
		        String[] phrases = text.split(" ");
        		return stream(searchTargets).flatMap(Supplier::get)
			               .filter(f -> stream(phrases).allMatch(phrase -> containsNoCase(f.getSearchText(), phrase)))
			               .sorted(by(Entry::getName))
			               .toList();
	        },
	        AutoCompletion.defaultStringConverter()
        );
        this.history = history;
        this.textField = textField;
        this.textField.setPrefWidth(450); // dictates the popup width
    }

    @Override
    protected AutoCompletePopup<Entry> buildPopup() {
        return new AutoCompletePopup<Entry>(){
            @Override
            protected Skin<AutoCompletePopup<Entry>> createDefaultSkin() {
                return new AutoCompletePopupSkin<>(this, 2) {
                    {
                        // set keys & allow typing
                        getSkinnable().addEventFilter(KeyEvent.KEY_PRESSED, e -> {
                            if (!ignoreEvent)
                                if (e.isControlDown() && (e.getCode()==KeyCode.UP || e.getCode()==KeyCode.DOWN)) {
                                    switch (e.getCode()) {
                                        case UP   : history.up(ConfigSearch.this); break;
                                        case DOWN : history.down(ConfigSearch.this); break;
                                        default   : break;
                                    }
                                } else if (e.getCode()==KeyCode.BACK_SPACE) {
	                                textField.deletePreviousChar();
	                                e.consume();
                                } else if (e.getCode()==KeyCode.DELETE) {
	                                textField.deleteNextChar();
	                                e.consume();
                                } else if (!e.getCode().isNavigationKey()) {
                                	// TODO: remove
                                    // We re-fire event on text field so we can type even though it
                                    // does not have focus. This causes event stack overflow, so we
                                    // defend against it with a boolean flag.
                                	if (!textField.isFocused()) {
	                                    ignoreEvent = true;
	                                    completionTarget.fireEvent(e);
	                                }
                                }
                            ignoreEvent = false;
                            // e.consume(); // may brake functionality
                        });
                        getNode().addEventFilter(KeyEvent.KEY_PRESSED, e -> {
                            if (!ignoreEvent)
                                if (e.isControlDown() && (e.getCode()==KeyCode.UP || e.getCode()==KeyCode.DOWN)) {
                                    switch (e.getCode()) {
                                        case UP   : history.up(ConfigSearch.this); break;
                                        case DOWN : history.down(ConfigSearch.this); break;
                                        default   : break;
                                    }
                                    e.consume();
                                } else if (e.isControlDown() && e.getCode()==KeyCode.A) {
                                    textField.selectAll();
                                    e.consume();
                                // } else if (e.getCode()==KeyCode.BACK_SPACE) {
                                ///    textField.deletePreviousChar(); // doesn't work here
                                //     e.consume();
                                } else if (e.getCode()==KeyCode.END) {
                                    if (e.isShiftDown()) textField.selectEnd();
                                    else textField.positionCaret(textField.getLength());
                                    e.consume();
                                } else if (e.getCode()==KeyCode.HOME) {
                                    if (e.isShiftDown()) textField.selectHome();
                                    else textField.positionCaret(0);
                                    e.consume();
                                } else if (e.getCode()==KeyCode.LEFT) {
                                    if (e.isControlDown()) textField.selectPreviousWord(); else textField.selectBackward();
                                    if (!e.isShiftDown()) textField.deselect();
                                    e.consume();
                                } else if (e.getCode()==KeyCode.RIGHT) {
                                    if (e.isControlDown()) textField.selectNextWord(); else textField.selectForward();
                                    if (!e.isShiftDown()) textField.deselect();
                                    e.consume();
                                } // else if (!e.getCode().isNavigationKey()) {}
                            ignoreEvent = false;
                            // e.consume(); // may brake functionality
                        });
                    }
                    @Override
                    protected ListCell<Entry> buildListViewCellFactory(ListView<Entry> view) {
                        return new EntryListCell();
                    }
                };
            }
        };
    }

    @Override
    protected void acceptSuggestion(Entry suggestion) {
		suggestion.run();
        history.add(this);
    }

    public static class History {
        private int history_at = 0;
        private final List<String> history = new ArrayList<>();

        private void up(ConfigSearch search) {
            if (history.isEmpty()) return;
            history_at = history_at==0 ? history.size()-1 : history_at-1;
            ((TextField)search.completionTarget).setText(history.get(history_at));
        }

        private void down(ConfigSearch search) {
            if (history.isEmpty()) return;
            history_at = history_at==history.size()-1 ? 0 : history_at+1;
            ((TextField)search.completionTarget).setText(history.get(history_at));
        }

        private void add(ConfigSearch search) {
            String last = history.isEmpty() ? null : history.get(history.size()-1);
            String curr = ((TextField)search.completionTarget).getText();
            boolean isDiff = last!=null && curr!=null && last.equalsIgnoreCase(curr);
            if (isDiff) {
                history.add(curr);
                history_at = history.size()-1;
            }
        }
    }
    public interface Entry extends Runnable {

	    static Entry of(Ƒ0<String> nameΛ, Runnable runΛ) {
		    return new ΛEntry(nameΛ, nameΛ, () -> null, runΛ);
	    }

	    static Entry of(Ƒ0<String> nameΛ, Ƒ0<String> infoΛ, Runnable runΛ) {
		    return new ΛEntry(nameΛ, infoΛ, () -> null, runΛ);
	    }

	    static Entry of(Ƒ0<String> nameΛ, Runnable runΛ, Ƒ0<Node> graphicsΛ) {
		    return new ΛEntry(nameΛ, nameΛ, graphicsΛ, runΛ);
	    }

	    static Entry of(Ƒ0<String> nameΛ, Ƒ0<String> infoΛ, Runnable runΛ, Ƒ0<Node> graphicsΛ) {
		    return new ΛEntry(nameΛ, infoΛ, graphicsΛ, runΛ);
	    }

    	static Entry of(Config<?> config) {
    	    return new ConfigEntry(config);
	    }

    	static Entry of(WidgetFactory f) {
    	    return new SimpleEntry(
                "Open widget " + f.nameGui(),
				"Open widget " + f.nameGui() + "\n\n" + "Opens the widget in new window.",
                () -> UiContext.launchComponent(f.name())
	        );
	    }

	    String getName();
	    default String getSearchText() {
	        return getName();
	    }
	    default String getInfo() {
		    return getName();
	    }
	    default Node getGraphics() {
	    	return null;
	    }

	    class ΛEntry implements Entry {
		    private final Ƒ0<String> nameΛ;
		    private final Ƒ0<String> infoΛ;
		    private final Ƒ0<Node> graphicsΛ;
		    private final Runnable runΛ;

		    public ΛEntry(Ƒ0<String> nameΛ, Ƒ0<String> infoΛ, Ƒ0<Node> graphicsΛ, Runnable runΛ) {
			    this.nameΛ = nameΛ;
			    this.infoΛ = infoΛ;
			    this.graphicsΛ = graphicsΛ;
			    this.runΛ = runΛ;
		    }

		    @Override
		    public String getName() {
			    return nameΛ.get();
		    }

		    @Override
		    public void run() {
			    runΛ.run();
		    }

		    @Override
		    public String getInfo() {
			    return infoΛ.get();
		    }

		    @Override
		    public Node getGraphics() {
			    return graphicsΛ.get();
		    }
	    }
	    class SimpleEntry implements Entry {
		    private final String name;
		    private final String info;
		    private final Runnable action;

		    SimpleEntry(String name, String info, Runnable action) {
			    this.name = name;
			    this.info = info;
			    this.action = action;
		    }

		    @Override
		    public String getName() {
			    return name;
		    }

		    @Override
		    public void run() {
			    if (action!=null) action.run();
		    }
	    }
	    class ConfigEntry implements Entry {
		    public final Config<?> config;
		    private boolean loaded = false;
		    private Node graphics;

		    ConfigEntry(Config<?> config) {
			    this.config = config;
		    }

		    @Override
		    public String getName() {
			    return config.getGroup() + "." + config.getGuiName();
		    }

		    @Override
		    public String getSearchText() {
			    return Action.class.isAssignableFrom(config.getType())
				           ? getName() + Action.class.cast(config).getKeys()
				           : getName();
		    }

		    @Override
		    public String getInfo() {
			    return getName() + "\n\n" + config.getInfo();
		    }

		    @Override
		    public Node getGraphics() {
			    if (loaded) return graphics;

			    if (config.getType()==Action.class && ((Action)config).hasKeysAssigned()) {
				    graphics = new Label(((Action)config).getKeys());
				    ((Label)graphics).setTextAlignment(TextAlignment.RIGHT);
			    } else {
				    graphics = config.getType()==Boolean.class || config.isTypeEnumerable()
					               ? ConfigField.create(config).getNode()
					               : null;
			    }

			    loaded = true;
			    return graphics;
		    }

		    @Override
		    public void run() {
			    if (Runnable.class.isAssignableFrom(config.getClass()))
				    ((Runnable)config).run();
			    else
			    if (Runnable.class.isAssignableFrom(config.getType()) && config.getValue() != null)
				    ((Runnable)config.getValue()).run();
		    }
	    }
    }
    private static class EntryListCell extends ListCell<Entry> {
        private final Label text = new Label();
        private final StackPane configNodeRoot = new StackPane();
        private final StackPane root = layStack(text,Pos.CENTER_LEFT, configNodeRoot,Pos.CENTER_RIGHT);
        private final Tooltip tooltip = appTooltip();

        public EntryListCell() {
            text.setTextAlignment(TextAlignment.LEFT);
            text.setTextOverrun(OverrunStyle.CENTER_ELLIPSIS);
            setMinPrefMaxSize(text,USE_COMPUTED_SIZE);
            text.prefWidthProperty().bind(root.widthProperty().subtract(configNodeRoot.widthProperty()).subtract(10));
            text.setMinWidth(200);
            text.maxWidthProperty().bind(root.widthProperty().subtract(100));
            text.setPadding(new Insets(5,0,0,10));
            Tooltip.install(root, tooltip);
        }

        @Override
        protected void updateItem(Entry item, boolean empty) {
            super.updateItem(item, empty);
            if (empty) {
                text.setText("");
                setGraphic(null);
            } else {
                if (getGraphic()!=root) setGraphic(root);

                tooltip.setText(item.getInfo());

                text.setText(item.getName());
	            Node node = item.getGraphics();
	            if (node instanceof HBox) ((HBox)node).setAlignment(Pos.CENTER_RIGHT);
	            if (node==null) configNodeRoot.getChildren().clear();
	            else {
	            	configNodeRoot.getChildren().setAll(node);
		            StackPane.setAlignment(node, Pos.CENTER_RIGHT);
	            }
            }
        }
    }
}

File: src/sp/it/pl/gui/objects/autocomplete/ConfigSearch.kt
package sp.it.pl.gui.objects.autocomplete

import javafx.geometry.Insets
import javafx.geometry.Pos.CENTER_LEFT
import javafx.geometry.Pos.CENTER_RIGHT
import javafx.scene.Node
import javafx.scene.control.Label
import javafx.scene.control.ListCell
import javafx.scene.control.ListView
import javafx.scene.control.OverrunStyle
import javafx.scene.control.TextField
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyEvent.KEY_PRESSED
import javafx.scene.layout.HBox
import javafx.scene.layout.StackPane
import javafx.scene.text.TextAlignment
import sp.it.pl.gui.itemnode.ConfigField
import sp.it.pl.gui.objects.autocomplete.ConfigSearch.Entry
import sp.it.pl.layout.widget.ComponentFactory
import sp.it.pl.main.APP
import sp.it.pl.main.appTooltip
import sp.it.pl.util.Util.containsNoCase
import sp.it.pl.util.access.minus
import sp.it.pl.util.action.Action
import sp.it.pl.util.conf.Config
import sp.it.pl.util.functional.setToOne
import sp.it.pl.util.graphics.install
import sp.it.pl.util.graphics.lay
import sp.it.pl.util.graphics.setMinPrefMaxSize
import sp.it.pl.util.graphics.stackPane
import sp.it.pl.util.reactive.onEventUp
import sp.it.pl.util.reactive.syncFrom
import java.util.ArrayList

class ConfigSearch: AutoCompletion<Entry> {
    private val textField: TextField
    private val history: History
    private var ignoreEvent = false

    constructor(textField: TextField, history: History = History(), entries: () -> Sequence<Entry>): super(
            textField,
            { text ->
                val phrases = text.split(" ").toList()
                entries()
                        .filter { phrases.all { phrase -> containsNoCase(it.searchText, phrase) } }
                        .sortedBy { it.name }
                        .toList()
            },
            AutoCompletion.defaultStringConverter()
    ) {
        this.textField = textField
        this.history = history
        this.textField.prefWidth = 450.0 // affects the popup width
    }

    override fun buildPopup(): AutoCompletePopup<Entry> {
        return object: AutoCompletePopup<Entry>() {

            override fun createDefaultSkin(): AutoCompletePopupSkin<Entry> {
                return object: AutoCompletePopupSkin<Entry>(this, 2) {

                    init {
                        // set keys & allow typing
                        skinnable.onEventUp(KEY_PRESSED) {
                            if (!ignoreEvent)
                                if (it.isControlDown && (it.code==KeyCode.UP || it.code==KeyCode.DOWN)) {
                                    when (it.code) {
                                        KeyCode.UP -> history.up(this@ConfigSearch)
                                        KeyCode.DOWN -> history.down(this@ConfigSearch)
                                        else -> {}
                                    }
                                } else if (it.code==KeyCode.BACK_SPACE) {
                                    textField.deletePreviousChar()
                                    it.consume()
                                } else if (it.code==KeyCode.DELETE) {
                                    textField.deleteNextChar()
                                    it.consume()
                                } else if (!it.code.isNavigationKey) {
                                    // We re-fire event on text field so we can type even though it
                                    // does not have focus. This causes event stack overflow, so we
                                    // defend with a flag.
                                    if (!textField.isFocused) {
                                        ignoreEvent = true
                                        completionTarget.fireEvent(it)
                                    }
                                }
                            ignoreEvent = false
                        }
                        node.onEventUp(KEY_PRESSED) {
                            if (!ignoreEvent)
                                if (it.isControlDown && (it.code==KeyCode.UP || it.code==KeyCode.DOWN)) {
                                    when (it.code) {
                                        KeyCode.UP -> history.up(this@ConfigSearch)
                                        KeyCode.DOWN -> history.down(this@ConfigSearch)
                                        else -> {}
                                    }
                                    it.consume()
                                } else if (it.isControlDown && it.code==KeyCode.A) {
                                    textField.selectAll()
                                    it.consume()
                                    // TODO:
                                    // else if (e.getCode()==KeyCode.BACK_SPACE) {
                                    //     textField.deletePreviousChar(); // doesn't work here
                                    //     e.consume();
                                } else if (it.code==KeyCode.END) {
                                    if (it.isShiftDown) textField.selectEnd() else textField.positionCaret(textField.length)
                                    it.consume()
                                } else if (it.code==KeyCode.HOME) {
                                    if (it.isShiftDown) textField.selectHome() else textField.positionCaret(0)
                                    it.consume()
                                } else if (it.code==KeyCode.LEFT) {
                                    if (it.isControlDown) textField.selectPreviousWord() else textField.selectBackward()
                                    if (!it.isShiftDown) textField.deselect()
                                    it.consume()
                                } else if (it.code==KeyCode.RIGHT) {
                                    if (it.isControlDown) textField.selectNextWord() else textField.selectForward()
                                    if (!it.isShiftDown) textField.deselect()
                                    it.consume()
                                }
                                // TODO: else if (!e.getCode().isNavigationKey()) {}
                            ignoreEvent = false
                        }
                    }

                    override fun buildListViewCellFactory(listView: ListView<Entry>) = EntryListCell()
                }
            }
        }
    }

    override fun acceptSuggestion(suggestion: Entry) {
        suggestion.run()
        history.add(this)
    }

    class History {
        private var historyIndex = 0
        private val history = ArrayList<String>()

        fun up(search: ConfigSearch) {
            if (history.isEmpty()) return
            historyIndex = if (historyIndex==0) history.size-1 else historyIndex-1
            search.completionTargetTyped.text = history[historyIndex]
        }

        fun down(search: ConfigSearch) {
            if (history.isEmpty()) return
            historyIndex = if (historyIndex==history.size-1) 0 else historyIndex+1
            search.completionTargetTyped.text = history[historyIndex]
        }

        fun add(search: ConfigSearch) {
            val last = if (history.isEmpty()) null else history[history.size-1]
            val curr = search.completionTargetTyped.text
            val isDiff = last!=null && curr!=null && !last.equals(curr, ignoreCase = true)
            if (isDiff) {
                history += curr
                historyIndex = history.size-1
            }
        }
    }

    @Suppress("NonAsciiCharacters", "ClassName")
    interface Entry: Runnable {
        val name: String
        val searchText: String get() = name
        val info: String get() = name
        val graphics: Node? get() = null

        class ΛEntry(nameΛ: () -> String, infoΛ: () -> String, searchTextΛ: () -> String, graphicsΛ: () -> Node, private val runΛ: () -> Unit): Entry {
            override val name = nameΛ()
            override val info = infoΛ()
            override val searchText = searchTextΛ()
            override val graphics = graphicsΛ()
            override fun run() = runΛ()
        }

        class SimpleEntry constructor(override val name: String, override val info: String, private val runΛ: () -> Unit): Entry {
            override fun run() = runΛ()
        }

        class ConfigEntry constructor(private val config: Config<*>): Entry {
            override val name = "${config.group}.${config.guiName}"
            override val searchText = if (config is Action) name+config.keys else name
            override val info = "$name\n\n${config.info}"
            override val graphics by lazy {
                when {
                    config is Action && config.hasKeysAssigned() -> {
                        Label(config.keys).apply {
                            textAlignment = TextAlignment.RIGHT
                        }
                    }
                    config.type==Boolean::class.java || config.isTypeEnumerable -> ConfigField.create(config).getNode()
                    else -> null
                }
            }

            @Suppress("UNCHECKED_CAST")
            override fun run() {
                val value = config.value
                when {
                    config is Runnable -> config.run()
                    value is Runnable -> value.run()
                    value is Boolean -> (config as Config<Boolean?>).setNapplyValue(!value)
                }
            }
        }

        companion object {

            fun of(nameΛ: () -> String, infoΛ: () -> String = { "" }, searchTextΛ: () -> String = nameΛ, graphicsΛ: () -> Node, runΛ: () -> Unit) = Entry.ΛEntry(nameΛ, infoΛ, searchTextΛ, graphicsΛ, runΛ)

            fun of(config: Config<*>) = Entry.ConfigEntry(config)

            fun of(f: ComponentFactory<*>): Entry {
                return SimpleEntry(
                        "Open widget ${f.nameGui()}",
                        "Open widget ${f.nameGui()}\n\nOpens the widget in new window.",
                        { APP.windowManager.launchComponent(f.create()) }
                )
            }
        }
    }

    private class EntryListCell: ListCell<Entry>() {
        private val text = Label()
        private val configNodeRoot = stackPane()
        private val root = stackPane {
            lay(CENTER_LEFT) += text
            lay(CENTER_RIGHT) += configNodeRoot
        }
        private val rootTooltip = appTooltip()

        init {
            text.textAlignment = TextAlignment.LEFT
            text.textOverrun = OverrunStyle.CENTER_ELLIPSIS
            text.setMinPrefMaxSize(USE_COMPUTED_SIZE)
            text.prefWidthProperty() syncFrom root.widthProperty()-configNodeRoot.widthProperty()-10
            text.minWidth = 200.0
            text.maxWidthProperty() syncFrom root.widthProperty()-100
            text.padding = Insets(5.0, 0.0, 0.0, 10.0)
            root install rootTooltip
        }

        override fun updateItem(item: Entry?, empty: Boolean) {
            super.updateItem(item, empty)

            if (empty || item==null) {
                text.text = ""
                graphic = null
            } else {
                graphic = root
                rootTooltip.text = item.info
                text.text = item.name

                val node = item.graphics
                if (node is HBox) node.alignment = CENTER_RIGHT
                if (node==null) {
                    configNodeRoot.children.clear()
                } else {
                    configNodeRoot.children setToOne node
                    StackPane.setAlignment(node, CENTER_RIGHT)
                }
            }
        }
    }
}